pkg/scheduler/actions/reserve/reserve.go

```diff
@@ -0,0 +1,249 @@
+		}
+		tasks := pendingTasks[job.UID]
+
+		glog.V(3).Infof("Trying to reserve resource for %d tasks of starving Job <%v/%v>",
+			tasks.Len(), job.Namespace, job.Name)
+
+		for !tasks.Empty() {
+			task := tasks.Pop().(*api.TaskInfo)
+
+			glog.V(3).Infof("There are <%d> nodes for Job <%v/%v>",
+				len(ssn.Nodes), job.Namespace, job.Name)
+
+			if len(job.NodesFitDelta) > 0 {
+				job.NodesFitDelta = make(api.NodeResourceMap)
+			}
+
+			// Going through all the nodes.
+			// When allocating for starving, node needs
+			// to be offered in consistent order
+			// If no node is available, task will not be allocated
+			// in any case
+			nodes := util.PredicateNodes(task, allNodes, predicateFn)
+			if len(nodes) == 0 {
+				break
+			}
+
+			taskAllocated := false
+			for _, node := range nodes {
+				// Allocate idle resource to the task.
+				if task.InitResreq.LessEqual(node.Idle) || toOverAllocate(ssn, node, threshold, task) {
+					if err := ssn.Allocate(task, node); err != nil {
+						glog.Errorf("Failed to bind task %v on %v in Session %v, err: %v",
+							task.UID, node.Name, ssn.UID, err)
+					}
+					taskAllocated = true
+					glog.V(3).Infof("Bond task <%v/%v> to node <%v>",
+						task.Namespace, task.Name, node.Name)
+					break
+				} else {
+					job.NodesFitDelta[node.Name] = node.Idle.Clone()
+					job.NodesFitDelta[node.Name].FitDelta(task.InitResreq)
+				}
+			}
+
+			// stop trying for this job if no (more) nodes available for this task
+			if !taskAllocated {
+				break
+			}
+
+			if ssn.JobReady(job) {
+				jobs.Push(job)
+				break
+			}
+		}
+
+		// Added Queue back until no job in Queue.
+		queues.Push(queue)
+	}
+}
+
+// ToOverAllocate check if it is okay to allocate resource on given node ignoring
+// running tasks
+func toOverAllocate(ssn *framework.Session, node *api.NodeInfo,
+	StarvationThreshold time.Duration, task *api.TaskInfo) bool {
+	job, found := ssn.Jobs[task.Job]
+	if !found {
+		return false
+	}
+	isJobStarving := isJobStarving(job, StarvationThreshold)
+	if !isJobStarving {
+		return false
+	}
+
+	// allow over-allocate for starving job
+	netResource := node.Allocatable.Clone()
+	for _, nodeTask := range node.Tasks {
+		if nodeTask.Status == api.Borrowing ||
+			nodeTask.Status == api.Allocated {
+			netResource.Sub(nodeTask.InitResreq)
+		}
+	}
+
+	// return true means when allocating for starving job, over-allocate resources ignoring
+	// resources being used on the node to prevent other non-startving job from taking the resources
+	return !task.InitResreq.LessEqual(node.Idle) &&
+		task.InitResreq.LessEqual(netResource)
+}
+
+func isJobStarving(jobInfo *api.JobInfo, starvationThreshold time.Duration) bool {
+	readiness := jobInfo.GetReadiness()
+	return readiness != api.Ready &&
+		readiness != api.ConditionallyReady &&
+		time.Since(jobInfo.CreationTimestamp.Time) >= starvationThreshold
+}
+
+func (alloc *reserveAction) getStarvationThreshold(threshold *time.Duration) error {
```

Comment:
Rather than passing a pointer, how about making `getStarvationThreshold` return `time.Duration`? 

```go
threshold, _ := time.ParseDuration(v1alpha1.DefaultStarvingTreshold)
alloc.arguments.GetTimeDuration(&threshold, "threshold")
return threshold
```

Reply:
fixed